userspace/libscap/scap_savefile.c

```diff
@@ -2968,3 +2977,107 @@ void scap_fseek(scap_t *handle, uint64_t off)
 			return;
 	}
 }
+
+int scap_reader_read_capture_plugin(scap_reader_t *r, uint8_t* buf, uint32_t len)
+{
+	uint32_t nread = 0;
+	ss_plugin_rc rc;
+
+	// If seek has legitimately been used backward, we first
+	// replay the last bytes of the previous read in the buffer.
+	if (r->m_seek_buffer_ptr != NULL && r->m_seek_buffer_len > 0)
+	{
+		nread = MIN(len, r->m_seek_buffer_len);
+		memcpy(buf, r->m_seek_buffer_ptr, nread);
+		r->m_seek_buffer_len -= nread;
+		r->m_seek_buffer_ptr += nread;
+		r->m_total_bytes_read += nread;
+		buf += nread;
+		len -= nread;
+	}
+	
+	// We want to loop on the plugin_read function because the plugin
+	// could return an SS_PLUGIN_TIMEOUT rc, which does not make sense
+	// for capture plugins. In that case, we sleep for a little and then
+	// resume the loop on plugin_read(). This still ensures that the read
+	// conceretly happens only once.
+	// This also skips the plugin_read() call if len is zero at this point.
+	while (len > 0) 
+	{
+		rc = r->m_capture_plugin->read(
+			r->m_capture_plugin->state,
+			r->m_capture_plugin->handle,
+			buf,
+			len,
+			&nread
+		);
+
+		switch (rc)
+		{
+			case SS_PLUGIN_SUCCESS:
+			case SS_PLUGIN_EOF:
+				// Update the total bytes read for tell/offset purposes
+				r->m_total_bytes_read += nread;
+
+				// Store the last few bytes of this read in our seek buffer,
+				// so that they can be re-used for seeking backwards.
+				r->m_seek_buffer_len = MIN(nread, READER_SEEK_BUF_SIZE);
+				memcpy(&r->m_seek_buffer[0], buf + nread - r->m_seek_buffer_len, r->m_seek_buffer_len);
+				r->m_seek_buffer_ptr = NULL;
+
+				return (int) nread;
+			case SS_PLUGIN_TIMEOUT:
+#ifdef _WIN32
+				Sleep((DWORD) BUFFER_EMPTY_WAIT_TIME_US_START / 1000);
+#else
+				usleep(BUFFER_EMPTY_WAIT_TIME_US_START);
+#endif
+				continue;

```

Comment:
Yeah, no good choice of action if a capture plugin returns timeout :) Still, can we at least move the ifdef somewhere? Like a header with a `static inline sleep_us(uint64_t usec)` with the ifdef inside? I believe we should encapsulate platform specifics in a small number of places, not sprawl them across the whole codebase.

Still, not going to NAK the PR over this (not the first random ifdef and probably not the last) so let me know your thoughts on this.

Reply:
I feel the same about this, I just thought that this PR was not the right place to address this too. I would prefer having some refactor/polishing-only PRs in the future.